[gui] Allow tabs to text in GUI

AlwinEsch opened the pull request:
## Description

This change introduces the support of tabs in texts, which are then correctly displayed in the GUI with up to 4 spaces.

The new text variable for this is `[TABS] tab amount [/TABS]`, or direct use of the `\t`.

The background is to have larger text blocks more structured, as individual spaces are more difficult to lay in lines in the correct vertical order.

<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

## Motivation and context

I once brought it into my binary addon system rework project and now thought of bringing it into master.
Really find it very useful in bigger things as a possibility.

Gosh. As a note to "xbmc/windows/GUIWindowSystemInfo.cpp", the thing really needs to be revised, because it doesn't really fit into international text styles :roll_eyes::smirk:.

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How has this been tested?

See test commit, but can also be direct inside strings.po.

<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->

## Screenshots (if appropriate):

Example before:
![Bildschirmfoto von 2021-10-03 13-21-38](https://user-images.githubusercontent.com/6879739/135751800-7caf79ab-a4fc-4602-b16e-d2cb06193577.png)
![Bildschirmfoto von 2021-10-03 13-27-14](https://user-images.githubusercontent.com/6879739/135751810-4af39f67-7994-4ca1-b767-f03aeb5afaf3.png)

Example after:
![Bildschirmfoto von 2021-10-03 13-20-56](https://user-images.githubusercontent.com/6879739/135751825-c826934f-8682-45a2-9748-2e8df80e3912.png)
![Bildschirmfoto von 2021-10-03 13-31-19](https://user-images.githubusercontent.com/6879739/135751843-0317d8ac-4f57-4465-a8a6-37bf44034b5a.png)
Note about first line, there it takes a bit longer, thats why not done for there :smile: 

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

lrusak commented:
I don't really understand the need for `[TABS]`? Can we not just use `\t` or is this for something special?

AlwinEsch commented:
Thought it is undesirable to take something like this directly in strings.po, since "\n" should also be used as "[CR]", for example.

The small positive with it is that it can be changed more easily with the number, e.g. `[TABS]5[/TABS]`.

But both are possible.

garbear commented:
I have mixed feelings on tabs. With fixed-width text, tabbing offers the same ability as spacing with less effort because you control the horizontal position (assuming you can control tabbing size, not the case in most apps, but it would be the case in Kodi).

However, for variable-width text, tabbing is unreliable. But inspecting the screenshots I only see tabbing in fixed-width text. Maybe we make fixed-width a requirement for tabbing?

AlwinEsch commented:
Yes, it has to be used with care and actually only if in strings.po's so that it is adapted to the respective languages.

The only thing I think this is very good is with regard to the TextViewer window from Kodi, as this contains larger things and is partly loaded from files (e.g. with my addon changelog request).

Allowing it definitely brings advantages, and tab use will soon be necessary when my project with Markdown support comes as text. For this purpose, the background has to be brought in a kind of reduced markdown (or as much as possible) in the text viewer.

So that something like this easily possible to add and not with hell of pain to have Kodi's string style :rofl::
![Screenshot from 2021-09-06 00-35-29](https://user-images.githubusercontent.com/6879739/132143598-e733bcc9-4248-4b2b-8b93-87b9093fcd17.png)

AlwinEsch commented:
Here is another background I wanted to bring it to:
![Bildschirmfoto vom 2021-01-28 21-50-39](https://user-images.githubusercontent.com/6879739/135875853-abdcd585-a666-4a13-90df-bae798156303.png)

Without having it in order, I found it harder to keep track of.

garbear commented:
@AlwinEsch concerning the last two pictures you posted,

I think that the first picture is a perfect example for the use of tabs. Code in dialog is an edge case but it can definitely use tabs.

On the other hand, the second shot shows tabs mixed with variable-width font. I think that block alignment is more useful than tab alignment in variable-width situations.

However, the cost of misalignment in variable-width tabbing is low, because even misaligned tabs can be more readable than no tabs, especially when the misalignment is patterned as in your second picture.

garbear commented on xbmc/guilib/GUIFontTTF.cpp:
```diff
@@ -1145,3 +1163,9 @@ void CGUIFontTTF::SetGlyphStrength(FT_GlyphSlot slot, int glyphStrength)
+
+float CGUIFontTTF::GetTabSpaceLength()
+{
+  const Character* c = GetCharacter(L'X', 0);
```
It's easier to process on fixed-width characters, and there's 2^32 total unicode points, so we just take the space hit and use `uint32_t` for all characters. In this case, both `'X'` ("X") and `L'X'` ("\00X") will expand to a padded "\00\00\00X".

However, thinking of type-correctness, I would perfer `'X'` here. Either way, we cast a unicode sub-page (8 bits or 16 bits) to a full unicode point implicitly. If we were using 16-bit `wchars` for processing, then we could avoid a cast using `L`. But this isn't the case, so we might as well simplify the constant. `L` just adds overhead for new programmers, impacting readability.

AlwinEsch commented:
If OK, I bring this in. About text viewer it is really good to have, with the others you are right makes it not direct sense, there better to separate parts into different controls (**<>Text view:</><>Value view</>**).

AlwinEsch committed:
[gui] Allow tabs to text in GUI

This change introduces the support of tabs in texts, which are then correctly displayed in the GUI with up to 4 spaces.

The new text variable for this is `[TABS] tab amount [/TABS]`, or direct use of the `\t`.

The background is to have larger text blocks more structured, as individual spaces are more difficult to lay in lines in the correct vertical order.

garbear reviewed (approved):
I don't see any more problems
